src/main.cpp

```diff
@@ -1,82 +1,92 @@
-
 // Copyright (c) 2015-16 Tom Deakin, Simon McIntosh-Smith,
 // University of Bristol HPC
 //
 // For full license terms please see the LICENSE file distributed with this
 // source code
 
-#include <iostream>
-#include <vector>
-#include <numeric>
-#include <cmath>
-#include <limits>
-#include <chrono>
 #include <algorithm>
-#include <iomanip>
+#include <chrono>
+#include <cmath>
 #include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <numeric>
+#include <vector>
 
 #define VERSION_STRING "5.0"
 
 #include "Stream.h"
+#include "StreamModels.h"
+#include "Unit.h"
 
-#if defined(CUDA)
-#include "CUDAStream.h"
-#elif defined(STD_DATA)
-#include "STDDataStream.h"
-#elif defined(STD_INDICES)
-#include "STDIndicesStream.h"
-#elif defined(STD_RANGES)
-#include "STDRangesStream.hpp"
-#elif defined(TBB)
-#include "TBBStream.hpp"
-#elif defined(THRUST)
-#include "ThrustStream.h"
-#elif defined(HIP)
-#include "HIPStream.h"
-#elif defined(HC)
-#include "HCStream.h"
-#elif defined(OCL)
-#include "OCLStream.h"
-#elif defined(USE_RAJA)
-#include "RAJAStream.hpp"
-#elif defined(KOKKOS)
-#include "KokkosStream.hpp"
-#elif defined(ACC)
-#include "ACCStream.h"
-#elif defined(SYCL)
-#include "SYCLStream.h"
-#elif defined(SYCL2020)
-#include "SYCLStream2020.h"
-#elif defined(OMP)
-#include "OMPStream.h"
-#elif defined(FUTHARK)
-#include "FutharkStream.h"
-#endif
+using namespace std;
 
 // Default size of 2^25
 int ARRAY_SIZE = 33554432;
 unsigned int num_times = 100;
 unsigned int deviceIndex = 0;
 bool use_float = false;
 bool output_as_csv = false;
-bool mibibytes = false;
-std::string csv_separator = ",";
-
-template <typename T>
-void check_solution(const unsigned int ntimes, std::vector<T>& a, std::vector<T>& b, std::vector<T>& c, T& sum);
+Unit unit = MegaByte;
+string csv_separator = ",";
 
-template <typename T>
-void run();
+// Benchmarks:
+constexpr size_t num_benchmarks = 6;
+array<char const*, num_benchmarks> labels = {"Copy", "Add", "Mul", "Triad", "Dot", "Nstream"};
+// Weights data moved by benchmark & therefore achieved BW:
+// bytes = weight * sizeof(T) * ARRAY_SIZE -> bw = bytes / dur
+array<size_t, num_benchmarks> weight = {/*Copy:*/ 2, /*Add:*/ 2, /*Mul:*/ 3, /*Triad:*/ 3, /*Dot:*/ 2, /*Nstream:*/ 4};
```

Comment:
It might be worth having a `struct Benchmark { std::string label; size_t weight;  }` and a `std::unordered_map` for the lookup or something.

Reply:
Have added the `struct`, but for now the `array` still seems fine.